File: go/03_stablecoin_mints_burns/main.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/dustin/go-humanize"
)

const (
	// Whale Alert API Key.
	// Replace this value or set the WHALE_ALERT_API_KEY environment variable.
	whaleAlertApiKey = "YOUR_API_KEY_HERE"

	apiBaseURL   = "https://leviathan.whale-alert.io"
	pollInterval = 3 * time.Second
	httpTimeout  = 15 * time.Second

	// Minimum USD threshold for alerts ($1,000 USD default)
	minUSDThreshold = 1_000.0
)

// Supported tracked stablecoins (lowercase symbol -> token name)
var stablecoins = map[string]string{
	"usdt":    "Tether",
	"usdc":    "USD Coin",
	"dai":     "Dai",
	"fei":     "Fei USD",
	"pax":     "Paxos Standard",
	"ust":     "TerraUSD",
	"busd":    "Binance USD",
	"eurt":    "Tether EUR",
	"gusd":    "Gemini Dollar",
	"husd":    "HUSD",
	"pusd":    "PUSD",
	"tusd":    "TrueUSD",
	"usdd":    "Decentralized USD",
	"usde":    "Ethena USDe",
	"usdg":    "Global Dollar",
	"usdh":    "USDH",
	"usdj":    "JUST Stablecoin",
	"usds":    "Sky Dollar",
	"usd1":    "USD1",
	"pyusd":   "PayPal USD",
	"rlusd":   "Ripple USD",
	"susds":   "Savings USDS",
	"pathusd": "PathUSD",
}

// HTTP client with configured timeout
var httpClient = &http.Client{
	Timeout: httpTimeout,
}

// StatusChain represents a blockchain network entry from GET /status
type StatusChain struct {
	Name    string   `json:"name"`
	Symbols []string `json:"symbols"`
	Height  uint64   `json:"height"`
}

// BlockchainStatus represents the response from GET /{blockchain}/status
type BlockchainStatus struct {
	EndHeight uint64 `json:"end_height"`
}

// TransactionsResponse represents the response from GET /{blockchain}/transactions
type TransactionsResponse struct {
	Transactions []Transaction `json:"transactions"`
	Next         string        `json:"next"`
}

// Transaction represents a single blockchain transaction
type Transaction struct {
	BlockHeight     uint64           `json:"height"`
	IndexInBlock    uint64           `json:"index_in_block"`
	Timestamp       int64            `json:"timestamp"`
	Hash            string           `json:"hash"`
	Fee             string           `json:"fee,omitempty"`
	FeeSymbol       string           `json:"fee_symbol,omitempty"`
	FeeSymbolPrice  float64          `json:"fee_symbol_price,omitempty"`
	SubTransactions []SubTransaction `json:"sub_transactions"`
}

// SubTransaction represents token/coin movement within a transaction
type SubTransaction struct {
	Symbol          string    `json:"symbol"`
	TransactionType string    `json:"transaction_type"`
	Price           float64   `json:"unit_price_usd"`
	Inputs          []Account `json:"inputs"`
	Outputs         []Account `json:"outputs"`
}

// Account represents a sender (input) or receiver (output)
type Account struct {
	Address     string `json:"address"`
	Amount      string `json:"amount"`
	Balance     string `json:"balance,omitempty"`
	Locked      string `json:"locked,omitempty"`
	IsFrozen    bool   `json:"is_frozen,omitempty"`
	Owner       string `json:"owner"`
	OwnerType   string `json:"owner_type"`
	AddressType string `json:"address_type"`
}

// Mutex for synchronized console output across blockchain workers
var printMu sync.Mutex

func main() {
	apiKeyFlag := flag.String("api-key", "", "Whale Alert API key (defaults to WHALE_ALERT_API_KEY env var)")
	flag.Parse()

	apiKey := getAPIKey(*apiKeyFlag)

	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer cancel()

	fmt.Printf("[%s] Stablecoin Mint, Burn, Freeze & Lock Monitor active (Threshold: $%s USD)...\n",
		time.Now().UTC().Format("2006-01-02 15:04:05"),
		humanize.CommafWithDigits(minUSDThreshold, 0),
	)

	// 1. Fetch available blockchains and symbols dynamically from the status endpoint
	allChains, err := fetchAvailableBlockchains(ctx, apiKey)
	if err != nil {
		log.Fatalf("Failed to fetch blockchain status: %v", err)
	}

	// 2. Filter blockchains that support our tracked stablecoins
	chainsToMonitor := getSupportedChains(allChains)
	if len(chainsToMonitor) == 0 {
		log.Fatalln("No blockchains found supporting the tracked stablecoins.")
	}

	chainNames := make([]string, 0, len(chainsToMonitor))
	for _, c := range chainsToMonitor {
		chainNames = append(chainNames, c.Name)
	}
	log.Printf("Discovered %d blockchains supporting stablecoins from /status: %s\n",
		len(chainsToMonitor), strings.Join(chainNames, ", "))

	// 3. Launch a monitoring goroutine for each discovered blockchain
	var wg sync.WaitGroup
	for _, chain := range chainsToMonitor {
		wg.Add(1)
		go monitorBlockchain(ctx, chain, apiKey, &wg)
	}

	// Wait for OS shutdown signal
	<-ctx.Done()
	log.Println("Cancellation signal received. Shutting down gracefully...")

	wg.Wait()
	log.Println("Stablecoin monitor stopped.")
}

// sleepContext pauses execution for d duration or until ctx is cancelled. Returns false if cancelled.
func sleepContext(ctx context.Context, d time.Duration) bool {
	select {
	case <-ctx.Done():
		return false
	case <-time.After(d):
		return true
	}
}

// getAPIKey resolves the API key from the -api-key flag, then the WHALE_ALERT_API_KEY
// environment variable, then the constant.
func getAPIKey(flagKey string) string {
	if flagKey != "" {
		return flagKey
	}
	if key := os.Getenv("WHALE_ALERT_API_KEY"); key != "" {
		return key
	}
	apiKey := whaleAlertApiKey
	if apiKey == "YOUR_API_KEY_HERE" {
		log.Fatal("Warning: WHALE_ALERT_API_KEY is not set. Please set the WHALE_ALERT_API_KEY environment variable, pass -api-key, or edit the constant in main.go.")
	}

	return apiKey
}

// fetchJSON executes an HTTP GET request, appends the API key, and decodes the JSON payload.
func fetchJSON(ctx context.Context, reqURL, apiKey string, target any) error {
	u, err := url.Parse(reqURL)
	if err != nil {
		return err
	}
	if apiKey != "" && u.Query().Get("api_key") == "" {
		q := u.Query()
		q.Set("api_key", apiKey)
		u.RawQuery = q.Encode()
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	if err != nil {
		return err
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		return err
	}
	defer func() {
		if err := resp.Body.Close(); err != nil {
			log.Printf("Error closing response body: %v", err)
		}
	}()

	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		bodyStr := strings.TrimSpace(string(bodyBytes))
		if bodyStr != "" {
			return fmt.Errorf("HTTP %d: %s", resp.StatusCode, bodyStr)
		}
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	return json.NewDecoder(resp.Body).Decode(target)
}

// fetchAvailableBlockchains retrieves all active blockchains and supported symbols from /status.
func fetchAvailableBlockchains(ctx context.Context, apiKey string) ([]StatusChain, error) {
	var chains []StatusChain
	err := fetchJSON(ctx, fmt.Sprintf("%s/status", apiBaseURL), apiKey, &chains)
	if err != nil {
		return nil, err
	}
	return chains, nil
}

// getSupportedChains filters blockchains that support at least one tracked stablecoin.
func getSupportedChains(allChains []StatusChain) []StatusChain {
	var matched []StatusChain
	for _, chain := range allChains {
		for _, sym := range chain.Symbols {
			if _, ok := stablecoins[strings.ToLower(sym)]; ok {
				matched = append(matched, chain)
				break
			}
		}
	}
	return matched
}

// monitorBlockchain queries and polls a single blockchain for stablecoin events.
func monitorBlockchain(ctx context.Context, chain StatusChain, apiKey string, wg *sync.WaitGroup) {
	defer wg.Done()

	startHeight := chain.Height
	// If height was not in status or is 0, fetch from /{blockchain}/status
	for startHeight == 0 {
		latestHeight, err := getLatestHeight(ctx, chain.Name, apiKey)
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			log.Printf("[%s] Failed to fetch latest block height: %v. Retrying in 5s...", chain.Name, err)
			if !sleepContext(ctx, 5*time.Second) {
				return
			}
			continue
		}
		startHeight = latestHeight
	}

	log.Printf("[%s] Monitoring from block #%d...\n", formatBlockchain(chain.Name), startHeight)

	// Query initial transactions from the start block height
	nextURL := fmt.Sprintf("%s/%s/transactions?start_height=%d", apiBaseURL, chain.Name, startHeight)

	// Keep polling periodically using the "next" URL cursor
	for ctx.Err() == nil {
		resp, err := fetchTransactions(ctx, nextURL, apiKey)
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			log.Printf("[%s] Error fetching transactions: %v. Retrying in 5s...", chain.Name, err)
			if !sleepContext(ctx, 5*time.Second) {
				return
			}
			continue
		}

		for _, tx := range resp.Transactions {
			checkTransaction(chain.Name, tx)
		}

		if resp.Next != "" {
			nextURL = resp.Next
		}

		if !sleepContext(ctx, pollInterval) {
			return
		}
	}
}

// getLatestHeight fetches the newest block height from the status endpoint.
func getLatestHeight(ctx context.Context, chain, apiKey string) (uint64, error) {
	var status BlockchainStatus
	err := fetchJSON(ctx, fmt.Sprintf("%s/%s/status", apiBaseURL, chain), apiKey, &status)
	if err != nil {
		return 0, err
	}
	return status.EndHeight, nil
}

// fetchTransactions retrieves transactions from the given URL and parses the response.
func fetchTransactions(ctx context.Context, fetchURL, apiKey string) (*TransactionsResponse, error) {
	if !strings.HasPrefix(fetchURL, "http") {
		fetchURL = apiBaseURL + fetchURL
	}

	var data TransactionsResponse
	err := fetchJSON(ctx, fetchURL, apiKey, &data)
	if err != nil {
		return nil, err
	}
	return &data, nil
}

// checkTransaction inspects a transaction's sub-transactions for supply and governance events:
// mints, burns, freezes, unfreezes, locks, and unlocks.
func checkTransaction(chain string, tx Transaction) {
	for _, sub := range tx.SubTransactions {
		symLower := strings.ToLower(sub.Symbol)

		// Step 1: Filter for tracked stablecoins
		if len(stablecoins) > 0 {
			if _, ok := stablecoins[symLower]; !ok {
				continue
			}
		}

		// Step 2: Filter for supported supply and governance event types
		switch sub.TransactionType {
		case "mint", "burn", "freeze", "unfreeze", "lock", "unlock":
			// Process supported supply and governance event types
		default:
			// Ignore standard transfers and unclassified movements
			continue
		}

		var amount float64

		// Step 3: Calculate token amount / address balance based on event type
		switch sub.TransactionType {
		case "mint":
			// Mint: creates new tokens at the output address(es)
			amount = sumAmounts(sub.Outputs)
		case "burn":
			// Burn: destroys tokens from the input address(es)
			amount = sumAmounts(sub.Inputs)
		case "lock", "unlock":
			// Lock / Unlock: tokens locked in or released from escrow
			amount = sumAmounts(sub.Inputs)
			if amount == 0 {
				amount = sumAmounts(sub.Outputs)
			}
		case "freeze":
			// Freeze: blacklists input address(es)
			amount = sumBalances(sub.Inputs)
		case "unfreeze":
			// Unfreeze: removes blacklist on output address(es)
			amount = sumBalances(sub.Outputs)
		}

		// If no positive amount could be calculated for transfer/mint/burn/lock/unlock, skip
		if sub.TransactionType != "freeze" && sub.TransactionType != "unfreeze" {
			if amount <= 0 {
				continue
			}
		}

		// Calculate USD valuation using the spot price provided by Whale Alert
		usdValue := amount * sub.Price

		// Filter out transactions below the configured minimum USD threshold to avoid noise
		if usdValue < minUSDThreshold {
			continue
		}

		// Format asset string and blockchain name
		assetStr := formatAsset(amount, sub.Symbol, usdValue)
		chainStr := formatBlockchain(chain)

		// Step 4: Resolve event-specific parties and labels
		var label1, party1, label2, party2 string
		amountLabel := "Asset"

		switch sub.TransactionType {
		case "mint":
			// Mints create new tokens to a recipient/treasury (destination only)
			label1, party1 = "To", getParty(sub.Outputs)
		case "burn":
			// Burns destroy tokens from a holder or issuer (source only)
			label1, party1 = "From", getParty(sub.Inputs)
		case "freeze":
			// Freezes blacklist a specific target address
			amountLabel = "Balance"
			label1, party1 = "Target", getParty(sub.Inputs)
		case "unfreeze":
			// Unfreezes restore transfer permissions to a target address
			amountLabel = "Balance"
			label1, party1 = "Target", getParty(sub.Outputs)
		case "lock":
			// Locks deposit tokens into a smart contract vault or bridge escrow
			label1, party1 = "From", getParty(sub.Inputs)
			label2, party2 = "Locked At", getParty(sub.Outputs)
		case "unlock":
			// Unlocks release escrowed tokens back into active circulation
			label1, party1 = "To", getParty(sub.Outputs)
		}

		// Step 5: Format and print event details to console (synchronized)
		printMu.Lock()
		fmt.Println()
		fmt.Printf("[STABLECOIN %s]\n", strings.ToUpper(sub.TransactionType))
		fmt.Printf("   %-12s %s\n", amountLabel+":", assetStr)
		fmt.Printf("   Blockchain:  %s\n", chainStr)
		if label1 != "" {
			fmt.Printf("   %-12s %s\n", label1+":", party1)
		}
		if label2 != "" {
			fmt.Printf("   %-12s %s\n", label2+":", party2)
		}
		if tx.Hash != "" {
			fmt.Printf("   Hash:        %s\n", tx.Hash)
		}
		printMu.Unlock()
	}
}

// sumAmounts sums parsed float amounts for a slice of accounts.
func sumAmounts(accounts []Account) float64 {
	var total float64
	for _, acc := range accounts {
		if amt, err := strconv.ParseFloat(acc.Amount, 64); err == nil && amt > 0 {
			total += amt
		}
	}
	return total
}

// sumBalances sums parsed float balances for a slice of accounts.
func sumBalances(accounts []Account) float64 {
	var total float64
	for _, acc := range accounts {
		if b, err := strconv.ParseFloat(acc.Balance, 64); err == nil && b > 0 {
			total += b
		}
	}
	return total
}

// formatAsset formats token amount, symbol, and USD valuation.
func formatAsset(amount float64, symbol string, valueUSD float64) string {
	symUpper := strings.ToUpper(symbol)
	if amount <= 0 {
		return fmt.Sprintf("0 %s", symUpper)
	}

	var amtStr string
	if amount == float64(int64(amount)) {
		amtStr = humanize.Comma(int64(amount))
	} else {
		amtStr = humanize.CommafWithDigits(amount, 2)
	}

	if valueUSD > 0 {
		return fmt.Sprintf("%s %s ($%s USD)", amtStr, symUpper, humanize.CommafWithDigits(valueUSD, 2))
	}
	return fmt.Sprintf("%s %s", amtStr, symUpper)
}

// formatBlockchain formats network names into standard title casing.
func formatBlockchain(chain string) string {
	if len(chain) == 0 {
		return "Unknown"
	}
	return strings.ToUpper(chain[:1]) + strings.ToLower(chain[1:])
}

// getParty resolves an account or list of accounts with formatted addresses and owner labels.
func getParty(accounts []Account) string {
	var parties []string
	seen := make(map[string]bool)

	for _, acc := range accounts {
		if acc.Address != "" && !isNullAddress(acc.Address) {
			if seen[acc.Address] {
				continue
			}
			seen[acc.Address] = true

			if acc.Owner != "" && !strings.EqualFold(acc.Owner, "unknown") {
				parties = append(parties, fmt.Sprintf("%s (%s)", formatShortAddress(acc.Address), acc.Owner))
			} else {
				parties = append(parties, formatShortAddress(acc.Address))
			}
		} else if acc.Owner != "" && !strings.EqualFold(acc.Owner, "unknown") && !seen[acc.Owner] {
			seen[acc.Owner] = true
			parties = append(parties, acc.Owner)
		}
	}

	if len(parties) > 0 {
		return strings.Join(parties, ", ")
	}
	return "Unknown"
}

// isNullAddress tests whether an address is a standard burn/null address.
func isNullAddress(addr string) bool {
	a := strings.ToLower(strings.TrimSpace(addr))
	return a == "" ||
		a == "null" ||
		a == "0x0" ||
		strings.HasPrefix(a, "0x0000000000000000000000000000000000000000") ||
		strings.HasPrefix(a, "0x000000000000000000000000000000000000dead") ||
		strings.HasPrefix(a, "11111111111111111111111111111111") ||
		a == "t9yd14nj9j7xab4dbgeix9h8unkkhxuwvb"
}

// formatShortAddress truncates long addresses for clean console output.
func formatShortAddress(addr string) string {
	addr = strings.TrimSpace(addr)
	if len(addr) > 12 {
		return fmt.Sprintf("%s...%s", addr[:6], addr[len(addr)-4:])
	}
	return addr
}

File: python/03_stablecoin_mints_burns/main.py
"""Whale Alert Quickstart: Stablecoin Mints, Burns, Freezes & Locks.

Discovers every blockchain that carries a tracked stablecoin, then polls each one
concurrently for supply and governance events: mints, burns, freezes, unfreezes,
locks, and unlocks above a configurable USD threshold.
"""

import argparse
import logging
import os
import signal
import sys
import threading
from datetime import datetime, timezone
from urllib.parse import urlencode, urlparse, urlunparse, parse_qs

import requests

# Whale Alert API Key.
# Replace this value or set the WHALE_ALERT_API_KEY environment variable.
WHALE_ALERT_API_KEY = "YOUR_API_KEY_HERE"

API_BASE_URL = "https://leviathan.whale-alert.io"
POLL_INTERVAL = 3  # seconds between polls
HTTP_TIMEOUT = 15  # seconds
RETRY_DELAY = 5  # seconds to wait after a failed request

# Minimum USD threshold for alerts ($1,000 USD default)
MIN_USD_THRESHOLD = 1_000.0

# Supported tracked stablecoins (lowercase symbol -> token name)
STABLECOINS = {
    "usdt": "Tether",
    "usdc": "USD Coin",
    "dai": "Dai",
    "fei": "Fei USD",
    "pax": "Paxos Standard",
    "ust": "TerraUSD",
    "busd": "Binance USD",
    "eurt": "Tether EUR",
    "gusd": "Gemini Dollar",
    "husd": "HUSD",
    "pusd": "PUSD",
    "tusd": "TrueUSD",
    "usdd": "Decentralized USD",
    "usde": "Ethena USDe",
    "usdg": "Global Dollar",
    "usdh": "USDH",
    "usdj": "JUST Stablecoin",
    "usds": "Sky Dollar",
    "usd1": "USD1",
    "pyusd": "PayPal USD",
    "rlusd": "Ripple USD",
    "susds": "Savings USDS",
    "pathusd": "PathUSD",
}

# Supply and governance event types this example reports on
TRACKED_EVENT_TYPES = ("mint", "burn", "freeze", "unfreeze", "lock", "unlock")

logging.basicConfig(format="%(asctime)s %(message)s", datefmt="%Y/%m/%d %H:%M:%S", level=logging.INFO)
log = logging.getLogger(__name__)

# Lock for synchronized console output across blockchain workers
print_lock = threading.Lock()

# Set on SIGINT / SIGTERM to stop every worker
shutdown = threading.Event()


def get_api_key(flag_key: str) -> str:
    """Resolve the API key from the --api-key flag, then WHALE_ALERT_API_KEY, then the constant."""
    if flag_key:
        return flag_key

    env_key = os.environ.get("WHALE_ALERT_API_KEY", "")
    if env_key:
        return env_key

    if WHALE_ALERT_API_KEY == "YOUR_API_KEY_HERE":
        log.error(
            "Warning: WHALE_ALERT_API_KEY is not set. Please set the WHALE_ALERT_API_KEY "
            "environment variable, pass --api-key, or edit the constant in main.py."
        )
        sys.exit(1)

    return WHALE_ALERT_API_KEY


def fetch_json(session: requests.Session, request_url: str, api_key: str):
    """Execute an HTTP GET request, append the API key, and decode the JSON payload."""
    parsed = urlparse(request_url)
    if api_key and "api_key" not in parse_qs(parsed.query):
        query = parse_qs(parsed.query)
        query["api_key"] = [api_key]
        parsed = parsed._replace(query=urlencode(query, doseq=True))

    response = session.get(urlunparse(parsed), timeout=HTTP_TIMEOUT)
    if response.status_code != 200:
        body = response.text.strip()
        raise RuntimeError(f"HTTP {response.status_code}: {body}" if body else f"HTTP {response.status_code}")

    return response.json()


def fetch_available_blockchains(session: requests.Session, api_key: str) -> list:
    """Retrieve all active blockchains and supported symbols from /status."""
    return fetch_json(session, f"{API_BASE_URL}/status", api_key)


def get_supported_chains(all_chains: list) -> list:
    """Filter blockchains that support at least one tracked stablecoin."""
    return [
        chain
        for chain in all_chains
        if any(str(symbol).lower() in STABLECOINS for symbol in chain.get("symbols") or [])
    ]


def get_latest_height(session: requests.Session, chain: str, api_key: str) -> int:
    """Fetch the newest block height from the status endpoint."""
    status = fetch_json(session, f"{API_BASE_URL}/{chain}/status", api_key)
    return int(status.get("end_height", 0))


def fetch_transactions(session: requests.Session, fetch_url: str, api_key: str) -> dict:
    """Retrieve transactions from the given URL and parse the response.

    The `next` cursor may come back as a relative path, so absolutize it first.
    """
    if not fetch_url.startswith(("http://", "https://")):
        if not fetch_url.startswith("/"):
            fetch_url = "/" + fetch_url
        fetch_url = API_BASE_URL + fetch_url

    return fetch_json(session, fetch_url, api_key)


def sum_field(accounts: list, field: str) -> float:
    """Sum a parsed float field ("amount" or "balance") across a list of accounts."""
    total = 0.0
    for account in accounts:
        try:
            value = float(account.get(field, 0) or 0)
        except (TypeError, ValueError):
            continue
        if value > 0:
            total += value
    return total


def format_asset(amount: float, symbol: str, value_usd: float) -> str:
    """Format token amount, symbol, and USD valuation."""
    symbol_upper = str(symbol).upper()
    if amount <= 0:
        return f"0 {symbol_upper}"

    # Whole numbers print without decimals; fractional amounts keep two.
    amount_str = f"{int(amount):,}" if amount == int(amount) else f"{amount:,.2f}"

    if value_usd > 0:
        return f"{amount_str} {symbol_upper} (${value_usd:,.2f} USD)"
    return f"{amount_str} {symbol_upper}"


def format_blockchain(chain: str) -> str:
    """Format network names into standard title casing."""
    return chain.capitalize() if chain else "Unknown"


def is_null_address(address: str) -> bool:
    """Test whether an address is a standard burn/null address."""
    addr = str(address).strip().lower()
    return (
        addr in ("", "null", "0x0", "t9yd14nj9j7xab4dbgeix9h8unkkhxuwvb")
        or addr.startswith("0x0000000000000000000000000000000000000000")
        or addr.startswith("0x000000000000000000000000000000000000dead")
        or addr.startswith("11111111111111111111111111111111")
    )


def format_short_address(address: str) -> str:
    """Truncate long addresses for clean console output."""
    address = str(address).strip()
    if len(address) > 12:
        return f"{address[:6]}...{address[-4:]}"
    return address


def get_party(accounts: list) -> str:
    """Resolve an account or list of accounts with formatted addresses and owner labels."""
    parties = []
    seen = set()

    for account in accounts:
        address = str(account.get("address", ""))
        owner = str(account.get("owner", ""))
        has_owner = bool(owner) and owner.lower() != "unknown"

        if address and not is_null_address(address):
            if address in seen:
                continue
            seen.add(address)
            if has_owner:
                parties.append(f"{format_short_address(address)} ({owner})")
            else:
                parties.append(format_short_address(address))
        elif has_owner and owner not in seen:
            # Burn/null addresses carry no useful address, so fall back to the entity label.
            seen.add(owner)
            parties.append(owner)

    return ", ".join(parties) if parties else "Unknown"


def check_transaction(chain: str, transaction: dict) -> None:
    """Inspect a transaction's sub-transactions for supply and governance events:
    mints, burns, freezes, unfreezes, locks, and unlocks.
    """
    for sub in transaction.get("sub_transactions") or []:
        symbol = str(sub.get("symbol", ""))
        tx_type = str(sub.get("transaction_type", ""))

        # Step 1: Filter for tracked stablecoins
        if STABLECOINS and symbol.lower() not in STABLECOINS:
            continue

        # Step 2: Filter for supported supply and governance event types
        # (ignores standard transfers and unclassified movements)
        if tx_type not in TRACKED_EVENT_TYPES:
            continue

        inputs = sub.get("inputs") or []
        outputs = sub.get("outputs") or []

        # Step 3: Calculate token amount / address balance based on event type
        if tx_type == "mint":
            # Mint: creates new tokens at the output address(es)
            amount = sum_field(outputs, "amount")
        elif tx_type == "burn":
            # Burn: destroys tokens from the input address(es)
            amount = sum_field(inputs, "amount")
        elif tx_type in ("lock", "unlock"):
            # Lock / Unlock: tokens locked in or released from escrow
            amount = sum_field(inputs, "amount") or sum_field(outputs, "amount")
        elif tx_type == "freeze":
            # Freeze: blacklists input address(es)
            amount = sum_field(inputs, "balance")
        else:
            # Unfreeze: removes blacklist on output address(es)
            amount = sum_field(outputs, "balance")

        # If no positive amount could be calculated for mint/burn/lock/unlock, skip
        if tx_type not in ("freeze", "unfreeze") and amount <= 0:
            continue

        # Calculate USD valuation using the spot price provided by Whale Alert
        try:
            unit_price = float(sub.get("unit_price_usd", 0) or 0)
        except (TypeError, ValueError):
            unit_price = 0.0
        usd_value = amount * unit_price

        # Filter out transactions below the configured minimum USD threshold to avoid noise
        if usd_value < MIN_USD_THRESHOLD:
            continue

        # Step 4: Resolve event-specific parties and labels
        amount_label = "Asset"
        label2 = party2 = ""

        if tx_type == "mint":
            # Mints create new tokens to a recipient/treasury (destination only)
            label1, party1 = "To", get_party(outputs)
        elif tx_type == "burn":
            # Burns destroy tokens from a holder or issuer (source only)
            label1, party1 = "From", get_party(inputs)
        elif tx_type == "freeze":
            # Freezes blacklist a specific target address
            amount_label = "Balance"
            label1, party1 = "Target", get_party(inputs)
        elif tx_type == "unfreeze":
            # Unfreezes restore transfer permissions to a target address
            amount_label = "Balance"
            label1, party1 = "Target", get_party(outputs)
        elif tx_type == "lock":
            # Locks deposit tokens into a smart contract vault or bridge escrow
            label1, party1 = "From", get_party(inputs)
            label2, party2 = "Locked At", get_party(outputs)
        else:
            # Unlocks release escrowed tokens back into active circulation
            label1, party1 = "To", get_party(outputs)

        # Step 5: Format and print event details to console (synchronized)
        with print_lock:
            print()
            print(f"[STABLECOIN {tx_type.upper()}]")
            print(f"   {amount_label + ':':<12} {format_asset(amount, symbol, usd_value)}")
            print(f"   {'Blockchain:':<12} {format_blockchain(chain)}")
            if label1:
                print(f"   {label1 + ':':<12} {party1}")
            if label2:
                print(f"   {label2 + ':':<12} {party2}")
            if transaction.get("hash"):
                print(f"   {'Hash:':<12} {transaction['hash']}")


def monitor_blockchain(chain: dict, api_key: str) -> None:
    """Query and poll a single blockchain for stablecoin events."""
    name = chain.get("name", "")

    # Each worker gets its own session: requests.Session is not designed for
    # concurrent use from multiple threads.
    session = requests.Session()

    start_height = int(chain.get("height", 0) or 0)

    # If height was not in status or is 0, fetch from /{blockchain}/status
    while start_height == 0 and not shutdown.is_set():
        try:
            start_height = get_latest_height(session, name, api_key)
        except (requests.RequestException, RuntimeError, ValueError) as exc:
            log.error("[%s] Failed to fetch latest block height: %s. Retrying in %ds...", name, exc, RETRY_DELAY)
            if shutdown.wait(RETRY_DELAY):
                return

    if shutdown.is_set():
        return

    log.info("[%s] Monitoring from block #%d...", format_blockchain(name), start_height)

    # Query initial transactions from the start block height
    next_url = f"{API_BASE_URL}/{name}/transactions?start_height={start_height}"

    # Keep polling periodically using the "next" URL cursor
    while not shutdown.is_set():
        try:
            response = fetch_transactions(session, next_url, api_key)
        except (requests.RequestException, RuntimeError, ValueError) as exc:
            log.error("[%s] Error fetching transactions: %s. Retrying in %ds...", name, exc, RETRY_DELAY)
            if shutdown.wait(RETRY_DELAY):
                return
            continue

        for transaction in response.get("transactions") or []:
            check_transaction(name, transaction)

        if response.get("next"):
            next_url = response["next"]

        if shutdown.wait(POLL_INTERVAL):
            return


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Monitor stablecoin mints, burns, freezes, unfreezes, locks, and unlocks across all supported chains."
    )
    parser.add_argument(
        "--api-key",
        default="",
        help="Whale Alert API key (defaults to WHALE_ALERT_API_KEY env var)",
    )
    args = parser.parse_args()

    api_key = get_api_key(args.api_key)

    def handle_signal(signum, _frame):
        log.info("Received signal %s. Shutting down gracefully...", signal.Signals(signum).name)
        shutdown.set()

    signal.signal(signal.SIGINT, handle_signal)
    if hasattr(signal, "SIGTERM"):
        signal.signal(signal.SIGTERM, handle_signal)

    now = datetime.now(tz=timezone.utc).strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{now}] Stablecoin Mint, Burn, Freeze & Lock Monitor active (Threshold: ${MIN_USD_THRESHOLD:,.0f} USD)...")

    # 1. Fetch available blockchains and symbols dynamically from the status endpoint
    try:
        all_chains = fetch_available_blockchains(requests.Session(), api_key)
    except (requests.RequestException, RuntimeError, ValueError) as exc:
        log.error("Failed to fetch blockchain status: %s", exc)
        sys.exit(1)

    # 2. Filter blockchains that support our tracked stablecoins
    chains_to_monitor = get_supported_chains(all_chains)
    if not chains_to_monitor:
        log.error("No blockchains found supporting the tracked stablecoins.")
        sys.exit(1)

    chain_names = [chain.get("name", "") for chain in chains_to_monitor]
    log.info(
        "Discovered %d blockchains supporting stablecoins from /status: %s",
        len(chains_to_monitor),
        ", ".join(chain_names),
    )

    # 3. Launch a monitoring thread for each discovered blockchain
    workers = []
    for chain in chains_to_monitor:
        worker = threading.Thread(target=monitor_blockchain, args=(chain, api_key), daemon=True)
        worker.start()
        workers.append(worker)

    # Wait for the shutdown signal. A timed wait keeps the main thread responsive
    # to Ctrl+C on Windows, where an untimed wait() cannot be interrupted.
    try:
        while not shutdown.wait(1):
            pass
    except KeyboardInterrupt:
        shutdown.set()

    log.info("Cancellation signal received. Shutting down gracefully...")

    for worker in workers:
        worker.join()

    log.info("Stablecoin monitor stopped.")


if __name__ == "__main__":
    main()
